src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -405,13 +471,97 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+        /// The locals are kept in a tree which captures the nesting of variables.
+        /// Each local is either a simple local (when its type is known) or a deconstruction local pending inference.
+        /// The caller is responsible for releasing the nested ArrayBuilders.
+        /// </summary>
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax closestTypeSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            SeparatedSyntaxList<VariableDeclarationSyntax> variables = node.Deconstruction.Variables;
+
+            // There are four cases for VariableDeclaration:
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. This could represent `(int x, ...)`
+
+            var localsBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                TypeSyntax typeSyntax = variable.Type ?? closestTypeSyntax;
+
+                DeconstructionVariable local;
+                if (variable.Deconstruction == null)
+                {
+                    local = new DeconstructionVariable(BindDeconstructionDeclarationVariable(variable, typeSyntax, diagnostics));
+                }
+                else
+                {
+                    local = new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, typeSyntax, diagnostics));
+                }
+
+                localsBuilder.Add(local);
+            }
+
+            return localsBuilder;
+        }
+
+        /// <summary>
+        /// Returns a BoundLocal when the type was explicit, otherwise returns a DeconstructionLocalPendingInference.
+        /// </summary>
+        private BoundExpression BindDeconstructionDeclarationVariable(VariableDeclarationSyntax node, TypeSyntax closestTypeSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction == null);
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+
+            var localSymbol = LocateDeclaredVariableSymbol(declarator, closestTypeSyntax);
+
+            // Check for variable declaration errors.
+            // Use the binder that owns the scope for the local because this (the current) binder
+            // might own nested scope.
+            bool hasErrors = localSymbol.Binder.ValidateDeclarationNameConflictsInScope(localSymbol, diagnostics);
+
+            if (node.Type != null)
```

Comment:
I am not sure why are we looking at ```node.Type``` here. I think we should just use ```closestTypeSyntax``` as the type syntax for the local and should remove this null check because we again looking at two different nodes. Should look just at one.